documentation/src/main/asciidoc/topics/ref_rest_cache_managers.adoc

```diff
@@ -204,3 +204,112 @@ include::json_examples/rest_cache_stats_response.json[]
 * `current_number_of_entries_in_memory` shows the total number of entries currently in all caches, excluding passivated entries.
 * `hit_ratio` provides the total percentage hit/(hit+miss) ratio for all caches.
 * `retrievals` shows the total number of `get()` operations.
+
+[id='rest_v2_backup_manager']
+= Backup {brandname} Cache Manager content
+Download a backup file (application/zip) containing all resources (Caches, Cache configurations, Counters, Proto schemas, Scripts)
+that are currently stored in the cache manager. The contents of the cluster backup file can then be restored
+via the link:#rest_v2_restore_manager[restore] action.
+
+[source,options="nowrap",subs=attributes+]
+----
+GET /rest/v2/cache-managers/{cacheManagerName}?action=backup
```

Comment:
@gustavonalle As a backup is non-idempotent does it make more sense for this to be a POST?

Reply:
Does it trigger a backup or simply re-downloads something that was already produced?

Reply:
It always initiates a new backup and downloads the result.

Reply:
According to the newly recent changes we were required to do, everything with side-effect should be a POST. 

But a POST downloading something is not "RESTful", and according to [1], there is a way to handle this: return a 201 code with a location header of the newly created resource or 202 if the resource was accepted but not ready.

[1] https://tools.ietf.org/html/rfc7231#section-6.3.2


I think we should make ```backup``` a proper resource:

1) Trigger the backup with a certain name and return ```202``` (Accepted)
```
POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

2)  Get the backup, returning it with ```200``` or ```202``` if it's not ready yet
```
GET /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

3) Delete a certain backup
```
DELETE /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

4) Restore a backup
```
POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}?action=restore
```

Reply:
I like this proposal a lot :+1:

For the CLI I can make it so that stages 1..3 appear as a single operation

Reply:
> 4. Restore a backup
> 
> ```
> POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}?action=restore
> ```

For this I have added very basic support for `multipart/form-data` that way the parameters used to filter the content from the backup file can be included as part of the payload.

Reply:
If I am not mistaken, we haven't been using ```multipart/form-data```, instead just use the JSON document as payload. 
FYI, this is how search handles it:

```POST /rest/v2/caches/{cacheName}?action=search```
```json
{
 "query":"from Entity where name:\"user1\"",
 "max_results":20,
 "offset":10
}
```

query params are part of the document

Reply:
The difference here is that we need to send both the `application/json` and the `application/zip` in a single request.

Reply:
fair enough, I was imagining that backups wouldn't travel via HTTP, but rather be triggered, restored and deleted from a known folder only.

Reply:
Unfortunately I don't believe that approach is flexible enough for Openshift and restoring backups to rolled out clusters. We [previously discussed](https://github.com/infinispan/infinispan/pull/8532#pullrequestreview-446504711) this and ways we can improve performance going forward, but I think for the initial PR at least the important thing to focus on is the REST API and BackupManager internals.

Reply:
I am thinking the scenario for large datasets, say GBs (or TBs) of data. 

Starting the CLI from another continent pointing to the cluster, and asking a backup will be very slow. A restore would involve transferring that amount of data again, all via http.
If the CLI is run in the same network as the cluster, it wouldn't be so bad, but it would still be faster if the server writes the backup to a certain file, and restore from that same place, without the data passing through REST. 

Another thing related to sizes, the rest server where the backup is triggered need to have space to hold the full cluster worth of data, otherwise it will have disk errors. The way I imagine backups working on cloud env is to attach a larger volume, with a certain proximity to the cluster, so this would imply the possibility for the user to choose the directory of destination.

WDYT?

Reply:
> The way I imagine backups working on cloud env is to attach a larger volume, with a certain proximity to the cluster, so this would imply the possibility for the user to choose the directory of destination.

I agree, this is the way to go. For typical Infinispan deployments there is no need for a large volume, so it's too restrictive to assume that  a server's volume would have the capacity to store content spread across the cluster.

I think that we should allow the following:

1. Allow users to specify an optional path when creating a backup. This determines both the working-dir and final location of the created backup file. If no path is specified, then we utilise the current location `ISPN_HOME/server/data/backup-manager/{backupName}/{backupName}.zip`

1. Restore via a path specified in the JSON payload. The file is processed from the locally available path with no file upload etc.

1. Restore via upload. Allows smaller backups to be restored without attaching additional volumes and simple Console user experience, i.e. upload backup via browser form.
    1. Users can optionally provide a custom working-dir path via JSON, if a larger volume is required for processing.

Reply:
+1 for items 1), 2)

Still in doubt about passing backups through HTTP though, but I am not ferociously against it :smile:

So assuming we decide to add later the path reference based backups as described above (in a follow up PR), it'd be wise to avoid disruptions to the REST API. I suppose
```
POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

Can be used for both cases right? If the backup is not there, then it should have a reference for a path.

Reply:
> So assuming we decide to add later the path reference based backups as described above (in a follow up PR), it'd be wise to avoid disruptions to the REST API. I suppose

I'm actually making the changes atm and it doesn't alter things too much, so it should be Ok to make it into this PR.

> Can be used for both cases right? If the backup is not there, then it should have a reference for a path.

:+1: I was thinking the JSON would be the following for referenced based backup and restore requests:

```json
{
  "backup-dir": "/some/path",
  "resources": {
     "caches": "[*]"
   }
}
```
The Content-Type of the restore request determines what course of action is taken:
* `application/zip` uploads the backup and processes all resource
* `application/json` reference based restore, reading backups from `backup-dir/{backupName}.zip` and only processing resources in the "resources" object if specified.
* `multipart/form-data` uploads the backup and processes the specified resources in the json. "backup-dir" still respected as the working directory if specified.

Reply:
Updated. I removed the option to upload via `application/zip` as netty allows files uploaded via `multipart/form-data` to be stored to disk, whereas that does not seem possible with `application/zip`. The latter is all that is required by html forms, so this shouldn't be an issue for the console.